In [2]:
import pandas as pd
import numpy as np
import os

### DataSet 1 - ***Predicting Churn for Bank Customers***

*Seleccione este dataset con la intención de analizar los datos y en base a estos ver como lo dice su nombre de predecir la deserción de clientes.*

In [3]:
nueva_ruta = 'C:\\Users\\REYES LUIS\\Downloads\\cursos\\Curso Data Science CoderHouse\\DataSets\\Bank Customer Churn Prediction'
os.chdir(nueva_ruta)

# Levanto el archivo 
df = pd.read_csv("Churn_Modelling.csv",sep = ',')
# Vamos algunos de sus datos
print(df.head())

   RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         790

In [4]:
# Veamos la cantidad de filas y columnas
df.shape

(10000, 14)

In [5]:
# Obtengamos mas información de las columnas del dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


In [6]:
df.isnull().sum()

RowNumber          0
CustomerId         0
Surname            0
CreditScore        0
Geography          0
Gender             0
Age                0
Tenure             0
Balance            0
NumOfProducts      0
HasCrCard          0
IsActiveMember     0
EstimatedSalary    0
Exited             0
dtype: int64

In [7]:
print(df.describe())

         RowNumber    CustomerId   CreditScore           Age        Tenure  \
count  10000.00000  1.000000e+04  10000.000000  10000.000000  10000.000000   
mean    5000.50000  1.569094e+07    650.528800     38.921800      5.012800   
std     2886.89568  7.193619e+04     96.653299     10.487806      2.892174   
min        1.00000  1.556570e+07    350.000000     18.000000      0.000000   
25%     2500.75000  1.562853e+07    584.000000     32.000000      3.000000   
50%     5000.50000  1.569074e+07    652.000000     37.000000      5.000000   
75%     7500.25000  1.575323e+07    718.000000     44.000000      7.000000   
max    10000.00000  1.581569e+07    850.000000     92.000000     10.000000   

             Balance  NumOfProducts    HasCrCard  IsActiveMember  \
count   10000.000000   10000.000000  10000.00000    10000.000000   
mean    76485.889288       1.530200      0.70550        0.515100   
std     62397.405202       0.581654      0.45584        0.499797   
min         0.000000     

*No contamos con una descripción del significado de cada columna pero su nombre es bastante intuitivo.*

*Los valores 0 en las diversas columnas no son significativos por lo que no requieren ser reemplazados.*

*Las columnas mas interesantes para realizar el analisis de deserción de clientes son:*

***CreditScore** => Puntaje crediticio, nos puede dar una pauta de si existe una relación entre el puntaje crediticio en aquellos clientes han dejado de serlo*            

***Age** => edad del cliente, con esta podemos entender mejor en que rango de edad se encuentran la mayor cantidad de clientes que han dejado de serlo*

***Balance** => balance de cuenta del cliente, nos ayudará a determinar si el balance de las cuentas de los clientes ha sido un factor de deserción*

***IsActiveMember** => Miembro activo, podemos saber si es un cliente frecuente o no*

***EstimatedSalary** => salario estimado, nos permitira saber mejor si el salario de los clientes tiene un impacto negativo o no en el nivel de deserción de estos*

***Exited** => salida, no informará si el cliente ya ha dejado de serlo*

*Si bien las otras variables presentan importancia, las antes mencionadas consideró como las mas importantes*

### DataSet 2 - ***Bank Marketing Analysis***

*Este dataset fue seleccionado con la intención de analizar la última campaña de marketing realizado por un banco, y en base a esta mejorar la estrategia para la siguiente campaña.*

In [10]:
nueva_ruta = 'C:\\Users\\REYES LUIS\\Downloads\\cursos\\Curso Data Science CoderHouse\\DataSets\\Bank Marketing Analysis'
os.chdir(nueva_ruta)

# Levanto el archivo 
df = pd.read_csv("bank.csv",sep = ',')
# Vamos algunos de sus datos
print(df.head())

   age         job  marital  education default  balance housing loan  contact  \
0   59      admin.  married  secondary      no     2343     yes   no  unknown   
1   56      admin.  married  secondary      no       45      no   no  unknown   
2   41  technician  married  secondary      no     1270     yes   no  unknown   
3   55    services  married  secondary      no     2476     yes   no  unknown   
4   54      admin.  married   tertiary      no      184      no   no  unknown   

   day month  duration  campaign  pdays  previous poutcome deposit  
0    5   may      1042         1     -1         0  unknown     yes  
1    5   may      1467         1     -1         0  unknown     yes  
2    5   may      1389         1     -1         0  unknown     yes  
3    5   may       579         1     -1         0  unknown     yes  
4    5   may       673         2     -1         0  unknown     yes  


In [11]:
# Veamos la cantidad de filas y columnas
df.shape

(11162, 17)

In [12]:
# Obtengamos mas información de las columnas del dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11162 entries, 0 to 11161
Data columns (total 17 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   age        11162 non-null  int64 
 1   job        11162 non-null  object
 2   marital    11162 non-null  object
 3   education  11162 non-null  object
 4   default    11162 non-null  object
 5   balance    11162 non-null  int64 
 6   housing    11162 non-null  object
 7   loan       11162 non-null  object
 8   contact    11162 non-null  object
 9   day        11162 non-null  int64 
 10  month      11162 non-null  object
 11  duration   11162 non-null  int64 
 12  campaign   11162 non-null  int64 
 13  pdays      11162 non-null  int64 
 14  previous   11162 non-null  int64 
 15  poutcome   11162 non-null  object
 16  deposit    11162 non-null  object
dtypes: int64(7), object(10)
memory usage: 1.4+ MB


In [13]:
df.isnull().sum()

age          0
job          0
marital      0
education    0
default      0
balance      0
housing      0
loan         0
contact      0
day          0
month        0
duration     0
campaign     0
pdays        0
previous     0
poutcome     0
deposit      0
dtype: int64

In [16]:
print(df.describe())

                age       balance           day      duration      campaign  \
count  11162.000000  11162.000000  11162.000000  11162.000000  11162.000000   
mean      41.231948   1528.538524     15.658036    371.993818      2.508421   
std       11.913369   3225.413326      8.420740    347.128386      2.722077   
min       18.000000  -6847.000000      1.000000      2.000000      1.000000   
25%       32.000000    122.000000      8.000000    138.000000      1.000000   
50%       39.000000    550.000000     15.000000    255.000000      2.000000   
75%       49.000000   1708.000000     22.000000    496.000000      3.000000   
max       95.000000  81204.000000     31.000000   3881.000000     63.000000   

              pdays      previous  
count  11162.000000  11162.000000  
mean      51.330407      0.832557  
std      108.758282      2.292007  
min       -1.000000      0.000000  
25%       -1.000000      0.000000  
50%       -1.000000      0.000000  
75%       20.750000      1.000000  


*Nuevamente tenemos un dataset que no cuenta con la descripción de sus columnas pero la mayoria son intuitivas. Las columnas Campaign, Pdays, Default, Poutcome y Previous nos cuesta discernir su papel en el dataset.*

*En este muestreo los datos encontrados no parecen todos validos, incluidos los campos con cero.*

*Las variables mas significativas en este dataset son:*

*La edad (**age**), el tipo de trabajo (**job**), el estado civil (**marital**), el nivel de educación (**education**), el balance (**balance**), si posee casa (**housing**), si ya tiene un préstamo (**loan**) y si tiene depositos (**deposit**)*

### DataSet 3 - ***Movie Recommendation System***

*Este es nuestro dataset favorito, ya que esta relacionado con algo que me gusta mucho que son las series y películas, como así también tiene la misma temática (no es el mismo dataset) que el proyecto final del curso de Data Analytics. La idea con este dataset es armar luego un sistema de recomendación de películas.*

In [17]:
nueva_ruta = 'C:\\Users\\REYES LUIS\\Downloads\\cursos\\Curso Data Science CoderHouse\\DataSets\\Movie Recommendation System'
os.chdir(nueva_ruta)

# Levanto el archivo 
df = pd.read_csv("tmdb_5000_movies.csv",sep = ',')
# Reordenemos el dataframe
df = df[['id', 'budget', 'genres', 'homepage', 'keywords', 'original_language', 'original_title', 'overview', 'popularity', 'production_companies', 'production_countries',
                'release_date', 'revenue', 'runtime', 'spoken_languages', 'status', 'tagline', 'title', 'vote_average', 'vote_count']]
# Importamos el segundo archivo
df_2 = pd.read_csv("tmdb_5000_credits.csv",sep = ',')

#Unamos los dos dataframes
df = pd.merge(df, df_2, on="id", how="outer")
# Vamos algunos de sus datos
print(df.head())


       id     budget                                             genres  \
0   19995  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1     285  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  206647  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3   49026  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4   49529  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage  \
0                   http://www.avatarmovie.com/   
1  http://disney.go.com/disneypictures/pirates/   
2   http://www.sonypictures.com/movies/spectre/   
3            http://www.thedarkknightrises.com/   
4          http://movies.disney.com/john-carter   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "nam

In [18]:
# Veamos la cantidad de filas y columnas
df.shape

(4803, 23)

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4803 entries, 0 to 4802
Data columns (total 23 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    4803 non-null   int64  
 1   budget                4803 non-null   int64  
 2   genres                4803 non-null   object 
 3   homepage              1712 non-null   object 
 4   keywords              4803 non-null   object 
 5   original_language     4803 non-null   object 
 6   original_title        4803 non-null   object 
 7   overview              4800 non-null   object 
 8   popularity            4803 non-null   float64
 9   production_companies  4803 non-null   object 
 10  production_countries  4803 non-null   object 
 11  release_date          4802 non-null   object 
 12  revenue               4803 non-null   int64  
 13  runtime               4801 non-null   float64
 14  spoken_languages      4803 non-null   object 
 15  status               

*Como podemos apreciar tenemos dos columnas con los mismos datos luego de la unión. Eliminemos una de ellas.*

***17  title_x**               4803 non-null   object*

***20  title_y**               4803 non-null   object*

In [20]:
df = df.drop(['title_y'], axis=1)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4803 entries, 0 to 4802
Data columns (total 22 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    4803 non-null   int64  
 1   budget                4803 non-null   int64  
 2   genres                4803 non-null   object 
 3   homepage              1712 non-null   object 
 4   keywords              4803 non-null   object 
 5   original_language     4803 non-null   object 
 6   original_title        4803 non-null   object 
 7   overview              4800 non-null   object 
 8   popularity            4803 non-null   float64
 9   production_companies  4803 non-null   object 
 10  production_countries  4803 non-null   object 
 11  release_date          4802 non-null   object 
 12  revenue               4803 non-null   int64  
 13  runtime               4801 non-null   float64
 14  spoken_languages      4803 non-null   object 
 15  status               

*Las variables mas significativas para llevar adelante el análisis y el posterior sistema de predicción son **budget, genres, keywords, popularity, vote_average y vote_count***

*Como se aprecia en el detalle anterior tenemos varios campos **null** por lo que vamos a controlar su información. Primero veamos cuantos hay.*

In [21]:
df.isnull().sum()

id                         0
budget                     0
genres                     0
homepage                3091
keywords                   0
original_language          0
original_title             0
overview                   3
popularity                 0
production_companies       0
production_countries       0
release_date               1
revenue                    0
runtime                    2
spoken_languages           0
status                     0
tagline                  844
title_x                    0
vote_average               0
vote_count                 0
cast                       0
crew                       0
dtype: int64

*Los campos **homepage, overview** y **tagline** si bien no tiene el 100% de los datos su faltante no nos afecta para el análisis.*

*Aunque los campos restantes con valores nulos son insignificantes, se puede completar para tener todos los datos*

*En el caso de **runtime** que es la duración de la película podemos usar el promedio general*

*Para **release_date** podiamos utilizar también un promedio, el primer dia de ese año u otro valor, pero también tenemos que ver que este campo sea un campo válido de fecha*

*Pero antes de continuar veamos con mas detalles los datos que tenemos*

In [22]:
print(df.describe())

                  id        budget   popularity       revenue      runtime  \
count    4803.000000  4.803000e+03  4803.000000  4.803000e+03  4801.000000   
mean    57165.484281  2.904504e+07    21.492301  8.226064e+07   106.875859   
std     88694.614033  4.072239e+07    31.816650  1.628571e+08    22.611935   
min         5.000000  0.000000e+00     0.000000  0.000000e+00     0.000000   
25%      9014.500000  7.900000e+05     4.668070  0.000000e+00    94.000000   
50%     14629.000000  1.500000e+07    12.921594  1.917000e+07   103.000000   
75%     58610.500000  4.000000e+07    28.313505  9.291719e+07   118.000000   
max    459488.000000  3.800000e+08   875.581305  2.787965e+09   338.000000   

       vote_average    vote_count  
count   4803.000000   4803.000000  
mean       6.092172    690.217989  
std        1.194612   1234.585891  
min        0.000000      0.000000  
25%        5.600000     54.000000  
50%        6.200000    235.000000  
75%        6.800000    737.000000  
max      

 *Houston we have a problem!!*

 *Bueno aparecieron nuevo problemas, tenemos campos válidos con valor cero lo que no nos podemos permitir*

 *Contemos cuantos tenemos de cada campo en cuestión*

In [31]:
nun_missing = (df[['budget','popularity','revenue','runtime','vote_average','vote_count']] == 0).sum()
print(nun_missing)


budget          1037
popularity         1
revenue         1427
runtime           35
vote_average      63
vote_count        62
dtype: int64


*Analicemos los cambios en los campos a corregir:*

*release_date => acá debemos convertir las fechas en un formato válido, y trabajar el null*

*runtime => reemplazaremos los valores null por el promedio*

*budget => el presupuesto de una película no puede ser cero, podemos reemplazar por el promedio los faltantes*

*popularity => este es un campo importante por lo que tenemos que reemplazar, y el promedio es un buen candidato para ello*

*revenue => la ganancia de la pelicula si bien puede ser cero es raro, aqui la desviacion estandar es un valor aceptable para reemplazar*

*runtime => la duración de las peliculas que esten en cero vamos a reemplazarlas por el promedio*

*vote_average => voto promedio, ya nos lo dice la palabra reemplazaremos por el promedio general de votos*

*vote_count => el conteo de votos es otro candidato a reemplazar los ceros por el promedio*

In [29]:
#Veamos el formato de las fechas
df.iloc[0:5, [11]]

release_date
0   2009-12-10
1   2007-05-19
2   2015-10-26
3   2012-07-16
4   2012-03-07